Add support to install collections from git repositories

s-hertel opened the pull request:
##### SUMMARY
Fixes #61680

This adds support for installing collections from git repositories. Use the format git+URI,version (or git@URI,version to use ssh authentication) instead of a collection name. The version is optional and should be a git tree-ish instead of a collection artifact version. The version defaults to HEAD.

A repo can contain multiple or an individual collection. You can append #/path/to/collection/ to the URI to install collections from a particular directory in the repository.

This also adds a new `type` option for collection requirements files (which may be one of 'galaxy', 'file', 'url', or 'git' to override implicitly determining the type from the `name` field).
```
collections:
  - name: git@github.com:my_org/private_collections.git#/path/to/collection,devel
  - name: https://github.com/ansible-collections/amazon.aws.git
    type: git
    version: 8102847014fd6e7a3233df9ea998ef4677b99248
```

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible-galaxy

s-hertel committed:
Enable installing collections from git repositories

s-hertel committed:
Add some tests

s-hertel committed:
Add some documentation

s-hertel committed:
Fix tests - update aliases and set the git config if necessary

s-hertel committed:
fix up documentation

add a changelog

s-hertel committed:
Skip Python 2.6

Fix warning

s-hertel committed:
Add a test to make sure recursive dependencies with different syntax are deduplicated

s-hertel commented:
cc @geerlingguy (since I can't request your review)

jborean93 commented on docs/docsite/rst/dev_guide/developing_collections.rst:
```diff
@@ -337,6 +337,19 @@ installs the collection in the first path defined in :ref:`COLLECTIONS_PATHS`, w
+
+   ansible-galaxy collection install git+https://github.com/org/repo.git,devel
+
+.. include:: ../shared_snippets/installing_collections_git_repo.txt
```
Won't this add another heading `Installing collections from a git repository from the shared segment?

jborean93 commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,83 @@
+       scm: git
+       version: devel
+
+Git repositories can be used for collection dependencies as well.
```
Might be a good idea to note that this isn't valid for actually published collections and just for local development.

jborean93 commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,83 @@
+
+   collections:
+     - name: https://github.com/organization/repo_name.git
+       scm: git
```
Should we add the `scm` key or have something like `type: git|file|url|galaxy`?

s-hertel replied:
Hm... I like `type:` better I think...

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -431,8 +495,15 @@ def from_path(b_path, force, parent=None):
 
         files = info.get('files_file', {}).get('files', {})
 
+        if fallback_metadata and not CollectionRequirement.artifact_info(b_path):
```
It's been a while since I last looked at this code. From my understanding `from_path` was used to identify collections which is why we unconditionally set `skip=True` so we never tried to install it again.

Am I understanding the code right where now `from_path` can be called against a cloned repo and that's why we may want to install it (not skip). If that's the case it seems like we may want to have the caller decide whether it's something to skip or not instead of using the is SCM like logic that's currently here.

s-hertel replied:
Looking for a better way of doing that.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -431,8 +495,15 @@ def from_path(b_path, force, parent=None):
+        if fallback_metadata and not CollectionRequirement.artifact_info(b_path):
+            skip = False
+        elif fallback_metadata:
+            skip = not force
```
Why is this when `not force`? It seems like it is saying skip a SCM collection that contains a `MANIFEST.json` only when `--force` is not set. It seems like we want to still install those collections and `--force` is only for reinstalling collections that are currently installed which is logic outside this function.

s-hertel replied:
I'm not sure what I was thinking here. This was to require --force to reinstall a collection from SCM when there was an installed artifact already. The if/elif should probably just change to:
```
if fallback_metadata:
    skip = False
else:
    skip = True
```

or pass a `skip` variable to this function and keep it separate from whether or not we're falling back to galaxy.yml - that seems better

jborean93 commented on lib/ansible/utils/galaxy.py:
```diff
@@ -0,0 +1,96 @@
+        except Exception as e:
+            ran = " ".join(cmd)
+            display.debug("ran %s:" % ran)
+            display.debug("\tstdout: " + to_text(stdout))
```
Would this even be set when `stdout`/`stderr` are only really set on the last call in the try block? It seems like if it's going to raise an exception then they will always be an empty string.

s-hertel replied:
Yeah, I've thought that before. This file is pretty much just code moved from lib/ansible/playbook/role/requirement.py and I decided not to refactor. I'll remove these couple lines though.

jborean93 commented on lib/ansible/utils/galaxy.py:
```diff
@@ -0,0 +1,96 @@
+    except (ValueError, OSError, IOError):
+        raise AnsibleError("could not find/use %s, it is required to continue with installing %s" % (scm, src))
+
+    tempdir = tempfile.mkdtemp(dir=C.DEFAULT_LOCAL_TMP)
```
Do we automatically cleanup `DEFAULT_LOCAL_TMP` on exit?

s-hertel replied:
yes, `_tempdir()` removes it

jborean93 commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,83 @@
+   # Install a collection from a local git repository
+   ansible-galaxy collection install git+file:///home/user/path/to/repo/.git
+
+In a ``requirements.yml`` file, you can also use the ``scm`` and ``version`` keys in addition to using the ``git+repo,version`` syntax for the collection name.
```
We should probably update https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/shared_snippets/installing_multiple_collections.txt to list the keys that are supported in a collection requirements and what they can be set to.

s-hertel replied:
Updated in d841de9c69f24d8fc433e6787c7071e68b7f28a2

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1054,6 +1168,54 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
+
+    for b_possible_collection in os.listdir(b_collection_path):
+        b_collection = os.path.join(b_collection_path, b_possible_collection)
+        b_galaxy = os.path.join(b_collection, b'galaxy.yml')
```
Currently the code is appending `galaxy.yml` to both files and folders so we should probably skip each entry if it's not a directory.

I just tested this against `ansible-galaxy collection install git+https://github.com/chocolatey/chocolatey-ansible` and it failed due to it not finding the `galaxy.yml` when I expected it to find it at `chocolatey/galaxy.yml`.

s-hertel replied:
That example fails for me too, but it seems like it's because the galaxy.yml is not valid for yaml.safe_loads rather than because the `galaxy.yml` can't be found.
```
ERROR! Failed to parse the galaxy.yml at '/home/user/.ansible/tmp/ansible-local-307040vg8x42it/tmpkah0w9xu/chocolatey-ansible/chocolatey/galaxy.yml' with the following error:
while constructing a mapping
  in "b'/home/user/.ansible/tmp/ansible-local-307040vg8x42it/tmpkah0w9xu/chocolatey-ansible/chocolatey/galaxy.yml'", line 12, column 10
found unhashable key
  in "b'/home/user/.ansible/tmp/ansible-local-307040vg8x42it/tmpkah0w9xu/chocolatey-ansible/chocolatey/galaxy.yml'", line 12, column 11
```

jborean93 replied:
Yea I came across that after explicitly setting the path but it still fails to find the galaxy.yml without it due to the current logic.

jborean93 replied:
Maybe we just need to add a note/warning to the docs saying that the galaxy.yml also needs to be valid for this to load. It might be overkill with all these warnings but I’m trying to cover use cases that people will try and then subsequently raise issues for where hopefully a note will stop some of that.

s-hertel replied:
Sounds good, I'll add one

webknjaz replied:
@s-hertel this is one of the hardcoded places that I mentioned in the comment below. Probably the last one...

s-hertel replied:
@webknjaz Oh, oops. I grepped, but somehow missed that one.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1054,6 +1168,54 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
+        if os.path.exists(b_galaxy):
+            reqs.append(CollectionRequirement.from_path(b_collection, force, parent, fallback_metadata=True))
+        else:
+            raise AnsibleError(err)
```
Should we only fail if `reqs` is an empty list, i.e. no `galaxy.yml` was found in any sub dirs? The way the docs are written is that this will only install dirs in SCM that contain the `galaxy.yml` one level deep. Right now it will fail unless either the root contains `galaxy.yml` or all the subdirs one level deep contain a `galaxy.yml`.

s-hertel replied:
Somehow I entirely misread that the first time. That definitely seems wrong, thanks for catching it.

jborean93 commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,83 @@
+Specifying the location to search for collections
+-------------------------------------------------
+
+If you have a different repository structure or only want to install a subset of collections, you can add a fragment to the end of your URI (before the optional comma-separated version) to indicate which path ansible-galaxy should inspect for ``galaxy.yml`` file(s).
```
I don't know whether we should add a note saying this is the path to the collection that contains the `galaxy.yml` and not the `galaxy.yml` itself. This tripped me up when testing and I only realised when stepping through the code and the reading this again.

jborean93 reviewed (changes requested):
Looks good, just a few docs questions and some behaviour around subdir checks

s-hertel committed:
Separate skip and whether or not to fallback to galaxy.yml in from_path

Remove a couple of uninformative debug messages in galaxy utils

s-hertel committed:
Some documentation fixes

Remove an extra header

don't use SCM dependencies for collection artifacts

Note that the fragment should be a path to a collection directory

s-hertel committed:
Only fail if no collections are located in a git repository

Add support for a 'type' key for collections in requirement.yml files.
Update the changelog and document the supported keys and allowed values for the type.

Add a note that the collection(s) in the repo must contain a galaxy.yml

s-hertel committed:
use a tuple

s-hertel committed:
Fix test collection requirements

jborean93 commented on docs/docsite/rst/shared_snippets/installing_multiple_collections.txt:
```diff
@@ -13,7 +13,11 @@ You can also setup a ``requirements.yml`` file to install multiple collections i
+
+The ``version`` key can take in the same range identifier format documented above. If you're installing a collection from a git repository instead of a built collection artifact, the ``version`` key refers to a git tree-ish.
+
+The ``type`` key can be set to ``galaxy``, ``url``, ``file``, and ``git``. If ``type`` is omitted, the ``name`` key is used to implicitly determine the source of the collection.
```
Don't we also support `hg`?

s-hertel replied:
No, `hg` works for roles. I initially included it and was told to limit it to `git` for collections.

jborean93 replied:
Fair enough, works for me.

jborean93 commented on lib/ansible/cli/galaxy.py:
```diff
@@ -504,6 +504,7 @@ def _parse_requirements_file(self, requirements_file, allow_old_format=True):
             - name: namespace.collection
               version: version identifier, multiple identifiers are separated by ','
               source: the URL or a predefined source name that relates to C.GALAXY_SERVER_LIST
+              type: git|file|url|galaxy
```
hg?

jborean93 commented on lib/ansible/utils/galaxy.py:
```diff
@@ -0,0 +1,94 @@
+        except Exception as e:
+            ran = " ".join(cmd)
+            display.debug("ran %s:" % ran)
+            raise AnsibleError("when executing %s: %s" % (ran, to_native(e)))
```
I think I saw someone talk about how this could expose someone's credentials if they are embedding it in the git URL Personally I think they should be using SSH keys instead but it might inadvertently expose info we don't want so worth a thought.

s-hertel replied:
I think credentials embedded in the git URL could be exposed multiple places, just as in the `Processing requirement collection ...` debug output (as well as already for roles installed from repositories). I don't know if I should change that in this PR or let it be a follow-up decision. Should I add a warning to the documentation?

jborean93 replied:
Yea a warning sounds good to me. I'm sure you are right where we output the URL specified in multiple places.

jborean93 reviewed (commented):
Looks good, only a question on the valid type entries where I think we should also document `hd`. Also just a quick question on logging the scm command we make on an error and whether we should or shouldn't do that.

jamescassell commented:
Does this support the `-g` flag like also for roles? (To keep the `.git` dir.)

s-hertel commented:
@jamescassell No, it doesn't. .git directories are [not included](https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/collection.py#L869) in collection artifact build process and this PR installs collections from the built artifact data using the galaxy.yml file(s).

s-hertel committed:
Update new tests and update assertions for new display message count

s-hertel committed:
Add a warning about embedding credentials in SCM URLs

s-hertel committed:
update another new test

s-hertel commented:
rebased again due to the test added in https://github.com/ansible/ansible/pull/69541

geerlingguy commented:
@s-hertel - I finally got time to fully test this and see if there were any weird edge cases, and I'm happy to report there didn't seem to be any.

There were two things that tripped me up (though not directly related to this PR's changes):

  1. For `role` installs, the `name` parameter reflects what will end up on the filesystem after an install, whereas for `collection` installs, the `name` can be one of galaxy namespace+collection name, or git path, or file tar.gz path (which is a little confusing coming from the roles perspective). For roles, the `src` was where you specified a filepath, git repo URL, or other source for the role. This is just a weird inconsistency between how collections and roles work, but since all collections _must_ be namespaced... I'm not sure if we can make any more consistency between roles/collections (roles don't have to be namespaced, something I like for playbook-local roles).
  2. If a collection doesn't have a valid `version` explicitly listed in its `galaxy.yml` (many don't currently) then installation fails. Also a number of collections I tested with had dependencies that didn't exist, so those installs failed (would've happened even installing from galaxy!).

Anyways, here's a test `requirements.yml` file I ended up with after testing a number of different permutations:

```yaml
---
roles:
  # Install a role from Ansible Galaxy.
  - name: geerlingguy.java
    version: 1.9.7

  # Install a role from GitHub, but specify a custom name for the local env.
  - name: homebrew
    src: https://github.com/geerlingguy/ansible-role-homebrew.git
    version: 3.1.0

collections:
  # Install a collection from Ansible Galaxy.
  - name: geerlingguy.php_roles
    version: 0.9.5
    source: https://galaxy.ansible.com

  # Install a collection from Git.
  - name: https://github.com/ansible-collections/community.kubernetes.git
    type: git
    version: 0.10.0

  # This doesn't work—depends on ansible.netcommon >=0.1.0 which doesn't exist.
  # - name: git+https://github.com/ansible-collections/community.crypto.git
  #   version: 0.1.0

  # This doesn't work—netcommon galaxy.yml doesn't contain a `version` key.
  # - name: git+https://github.com/ansible-collections/ansible.netcommon.git
  #   version: 0.0.2

  # Install from git+https (don't have to specify 'type'), at commit hash.
  - name: git+https://github.com/ansible-collections/community.zabbix.git
    version: b247fcf
```

s-hertel commented:
@geerlingguy Thanks for testing this.

> For role installs, the name parameter reflects what will end up on the filesystem after an install,

I thought about how this feature might work for collections if `src` was supported and `name` was allowed as a destination instead, but content in a collection could need to be edited to remain functional (some plugins refer to their own namespace, for example).

> If a collection doesn't have a valid version explicitly listed in its galaxy.yml (many don't currently) then installation fails.

Just to clarify, version is a [mandatory metadata key](https://docs.ansible.com/ansible/devel/dev_guide/collections_galaxy_meta.html#collections-galaxy-meta) and it needs to be valid YAML, but it doesn't need to be compatible with semantic versioning until built+published. It could just be a variable, for instance.

geerlingguy commented:
> it doesn't need to be compatible with semantic versioning until built+published

Ah, perfect! So if you build your collection and use a placeholder in the git repository (e.g. `version: '{{ build_version_here }}'`), that should work?

s-hertel commented:
@geerlingguy Yes!

jborean93 commented:
It's the same requirement for `ansible-galaxy collection build`, the content of the key is only validated once published, the build doesn't care about the value, only that it's there and set.

samdoran commented on docs/docsite/rst/shared_snippets/installing_multiple_collections.txt:
```diff
@@ -13,7 +13,11 @@ You can also setup a ``requirements.yml`` file to install multiple collections i
      source: 'The Galaxy URL to pull the collection from (default: ``--api-server`` from cmdline)'
 
-The ``version`` key can take in the same range identifier format documented above.
+The support keys for collection requirement entries are ``name``, ``version``, ``source``, and ``type``.
```
```suggestion
The supported keys for collection requirement entries are ``name``, ``version``, ``source``, and ``type``.
```

samdoran commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1008,6 +1064,47 @@ def reset_stat(tarinfo):
+        if file_info['name'] == '.':
+            continue
+
+        src_file = os.path.join(b_collection_path, to_bytes(file_info['name']))
+        dest_file = os.path.join(b_collection_output, to_bytes(file_info['name']))
```
This may not be necessary since it may already be encoded on line 1171.
```suggestion
        src_file = os.path.join(b_collection_path, to_bytes(file_info['name'], errors='surrogate_or_strict'))
        dest_file = os.path.join(b_collection_output, to_bytes(file_info['name'], errors='surrogate_or_strict'))
```

samdoran commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1167,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
+    else:
+        b_collection_path = os.path.join(b_temp_path, b_repo_root)
+
+    b_galaxy_path = os.path.join(b_collection_path, b'galaxy.yml')
```
Not a big deal now, but I can see hard coding `galaxy.yml` in places making it slightly harder in the future to support `galaxy.yaml` when someone inevitably asks for it to be supported.

s-hertel replied:
I went ahead and added a function in 0e703ca44050eebd7468d29043dfc4c09df3f151 to get the galaxy.yml/galaxy.yaml path and updated the changelog to note that both are supported now.

samdoran replied:
Very nice!

samdoran reviewed (approved):
This looks really great overall. Just a few minor things.

felixfontein commented on lib/ansible/cli/galaxy.py:
```diff
@@ -705,12 +710,13 @@ def _require_one_of_collections_requirements(self, collections, requirements_fil
                 if os.path.isfile(to_bytes(collection_input, errors='surrogate_or_strict')) or \
-                        urlparse(collection_input).scheme.lower() in ['http', 'https']:
+                        urlparse(collection_input).scheme.lower() in ['http', 'https'] or \
+                        collection_input.startswith('git+'):
```
```suggestion
                        collection_input.startswith('git+') or \:
                        collection_input.startswith('git@'):
```
Otherwise `ansible-galaxy collection install git@github.com:ansible-collections/community.azure.git` does not work:
```
ERROR! - command /usr/bin/git clone git@github.com git@github.com failed in directory /.../tmp/ansible-local-329971lp8my187/tmpi20ny9am (rc=128) - fatal: repository 'git@github.com' does not exist
```
(It does work when prepending `git+`. But just using the clone URL github spits out is nicer :) )

s-hertel replied:
Thanks, glad you spotted that

webknjaz commented on docs/docsite/rst/dev_guide/developing_collections.rst:
```diff
@@ -339,6 +339,19 @@ installs the collection in the first path defined in :ref:`COLLECTIONS_PATHS`, w
 
+.. _collections_scm_install:
+
+Installing collections from a git repository
+--------------------------------------------
```
You can't install more than one collection from a repo, can you?
```suggestion
Installing a collection from a git repository
---------------------------------------------
```

s-hertel replied:
Yes, you can. Either by having multiple directories with galaxy.yml files in the root of the repo or by adding a fragment to the URI.

webknjaz commented on docs/docsite/rst/shared_snippets/installing_multiple_collections.txt:
```diff
@@ -13,7 +13,11 @@ You can also setup a ``requirements.yml`` file to install multiple collections i
-The ``version`` key can take in the same range identifier format documented above.
+The support keys for collection requirement entries are ``name``, ``version``, ``source``, and ``type``.
+
+The ``version`` key can take in the same range identifier format documented above. If you're installing a collection from a git repository instead of a built collection artifact, the ``version`` key refers to a git tree-ish.
```
Looks like using commit-ish here is a bit more accurate: https://stackoverflow.com/a/23303550/595220.

Also, it's probably worth adding a link to https://git-scm.com/docs/gitglossary#def_commit-ish.

s-hertel replied:
TIL. :+1:

webknjaz commented on lib/ansible/utils/galaxy.py:
```diff
@@ -0,0 +1,94 @@
+# (c) 2014 Michael DeHaan, <michael@ansible.com>
```
@s-hertel this copyright statement looks inaccurate. Is it copied over as is?

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1119,6 +1298,32 @@ def _get_collection_info(dep_map, existing_collections, collection, requirement,
 
 
+def parse_scm(collection, version):
+    if ',' in collection:
```
@s-hertel why use comma? It's quite unconventional. Is `@` already taken by something else?

s-hertel replied:
It's the syntax supported by roles and was just there for similarity.

webknjaz replied:
Ah, I see. I haven't used it so it looks weird based on the experience with other ecosystems. But in this case, consistency makes sense.

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1119,6 +1298,32 @@ def _get_collection_info(dep_map, existing_collections, collection, requirement,
     dep_map[to_text(collection_info)] = collection_info
 
 
+def parse_scm(collection, version):
```
Why didn't you use urllib split?

s-hertel replied:
Mostly for similarity with roles. If you have a recommendation for a better way I'd update this.

webknjaz replied:
I didn't know the detail about roles. It just looked like a lot of manual processing which may be error-prone.

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1167,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
+
+def get_collection_info_from_req(dep_map, collection):
+    collection_name = to_text(collection)
+    if collection_name in dep_map:
```
Looks like applying a guard-check style would make it simpler and reduce nesting:
```suggestion
    if collection_name not in dep_map:
        return collection
```

s-hertel replied:
I'm going to keep this as-is unless you feel very strongly about it. It's just my preference, but I generally like to have one return statement for readability. This would let me remove the `else` line but is mostly stylistic.

webknjaz replied:
okay, this function is not that complex atm

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1167,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
 
     b_tar_path = None
-    if os.path.isfile(to_bytes(collection, errors='surrogate_or_strict')):
+    if req_type == 'file' or (not req_type and os.path.isfile(to_bytes(collection, errors='surrogate_or_strict'))):
```
It'd be great for readability if this clause was in a named var like
```suggestion
    IS_FILE = (
        req_type == 'file' or
        (not req_type and os.path.isfile(to_bytes(collection, errors='surrogate_or_strict')))
    )
    if IS_FILE:
```

s-hertel replied:
Done for all three like this. I made the vars lowercase though.

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1167,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
+
+    b_galaxy_path = os.path.join(b_collection_path, b'galaxy.yml')
+
+    err = "{0} appears to be an SCM collection source, but the required galaxy.yml was not found. " \
+          "Append #path/to/collection/ to your URI (before the comma separated version, if one is specified) " \
+          "to point to a directory containing the galaxy.yml or directories of collections".format(collection)
```
Usually, I discourage using EOL escaping because things like
```python
st = 'some words' \

'separate'
```
are valid. And `'separate'` in this example wouldn't be a part of `st`.  One accidentally added or removed empty line makes a lot of difference.

s-hertel replied:
good point

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1167,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
+    :param b_temp_path: The temporary path to the archive of a collection
+    :param force: Whether to overwrite an existing collection or fail
+    :param parent: The name of the parent collection
+    :return: List of CollectionRequirement objects
```
```suggestion
    :raises AnsibleError: if nothing found
    :return: List of CollectionRequirement objects
    :rtype: list
```

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1167,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
 
 
+def _collections_from_scm(collection, requirement, b_temp_path, force, parent=None):
+    """
+    Returns a list of collections found in the repo. If there is a galaxy.yml in the collection then just return
+    the specific collection. Otherwise, check each top-level directory for a galaxy.yml.
```
If it's meant to be compliant w/ PEP257, it should be:
```suggestion
    """Return a list of collections found in the repo.

    If there is a galaxy.yml in the collection then just return
    the specific collection. Otherwise, check each top-level
    directory for a galaxy.yml.
```

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1008,6 +1064,47 @@ def reset_stat(tarinfo):
+        src_file = os.path.join(b_collection_path, to_bytes(file_info['name']))
+        dest_file = os.path.join(b_collection_output, to_bytes(file_info['name']))
+
+        if any([src_file.startswith(directory) for directory in base_directories]):
```
Creating in-memory throw-away lists can be avoided when using functions that support iterables. Besides saving some memory, it's also faster because the generator is not necessarily looped over completely:
```suggestion
        if any(src_file.startswith(directory) for directory in base_directories):
```

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1008,6 +1064,47 @@ def reset_stat(tarinfo):
+    # Write contents to the files
+    for name, b in [('MANIFEST.json', collection_manifest_json), ('FILES.json', files_manifest_json)]:
+        b_path = os.path.join(b_collection_output, to_bytes(name, errors='surrogate_or_strict'))
+        b_io = BytesIO(b)
+        with open(b_path, 'wb') as file_obj:
+            shutil.copyfileobj(b_io, file_obj)
```
BytesIO is also a file-like object, so use it with CM. Otherwise, you'll end up with the `ResourceWarning: unclosed file <BytesIO(...)>...` (it's not printed out w/o smth like `-Xdev` but it's there):
```suggestion
        with open(b_path, 'wb') as file_obj, BytesIO(b) as b_io:
            shutil.copyfileobj(b_io, file_obj)
```

webknjaz replied:
(I don't remember if Py 2.7 can do two CMs in one `with`, if not — change it to two nested blocks)

s-hertel replied:
Looks like it works on Py 2.7.

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -968,6 +1023,7 @@ def _build_manifest(namespace, name, version, authors, readme, tags, description
 
 
 def _build_collection_tar(b_collection_path, b_tar_path, collection_manifest, file_manifest):
+    """ Build a tar.gz collection artifact from the manifest data. """
```
```suggestion
    """Build a tar.gz collection artifact from the manifest data."""
```

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1008,6 +1064,47 @@ def reset_stat(tarinfo):
 
 
+def _build_collection_dir(b_collection_path, b_collection_output, collection_manifest, file_manifest):
+    """ Build a collection directory from the manifest data. This should follow the same pattern as _build_collection_tar. """
```
```suggestion
    """Build a collection directory from the manifest data.

    This should follow the same pattern as _build_collection_tar.
    """
```

s-hertel committed:
Update with sdoran and felixfontein's suggestions

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -235,6 +283,30 @@ def install(self, path, b_temp_path):
             raise
 
+    def install_scm(self, b_collection_output_path):
+        """
+        Generates the Ansible collection artifact data from a galaxy.yml and installs the artifact to a directory.
+        This should follow the same pattern as build_collection, but instead of creating an artifact, install it.
+        :param b_collection_output_path: The installation directory for the collection artifact.
+        """
```
```suggestion
        """Install collection from source control into given dir.

        Generates the Ansible collection artifact data from a galaxy.yml and installs
         the artifact to a directory.
        This should follow the same pattern as build_collection, but instead of creating
        an artifact, install it.

        :param bytes b_collection_output_path: The installation directory for the collection artifact.

        :raises AnsibleError: If no collection metadata found.
        """
```

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -140,6 +142,46 @@ def dependencies(self):
+    @staticmethod
+    def collection_info(b_path, fallback_metadata=False):
+        info = CollectionRequirement.artifact_info(b_path)
+        if not info and fallback_metadata:
+            info = CollectionRequirement.galaxy_metadata(b_path)
+
+        return info
```
Less assigns?
```suggestion
        if info or not fallback_metadata:
            return info

        return CollectionRequirement.galaxy_metadata(b_path)
```

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1168,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
+    name, version, path, fragment = parse_scm(collection, requirement)
+    b_repo_root = to_bytes(name, errors='surrogate_or_strict')
+
+    if fragment:
+        b_fragment = to_bytes(fragment, errors='surrogate_or_strict')
+        b_collection_path = os.path.join(b_temp_path, b_repo_root, b_fragment)
+    else:
+        b_collection_path = os.path.join(b_temp_path, b_repo_root)
```
```suggestion
    b_collection_path = os.path.join(b_temp_path, b_repo_root)
    if fragment:
        b_fragment = to_bytes(fragment, errors='surrogate_or_strict')
        b_collection_path = os.path.join(b_collection_path, b_fragment)
```

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1168,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
+
+        # Ignore requirement if it is set (it must follow semantic versioning, unlike a git version, which is any tree-ish)
+        # If the requirement was the only place version was set, requirement == version at this point
+        if requirement not in ("*", "",) and requirement != version:
```
It's semantically better to use a set:
```suggestion
        if requirement not in {"*", ""} and requirement != version:
```

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1070,46 +1168,128 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
-            collection_info = req
+            validate_collection_name(collection)
+
+            display.vvvv("Collection requirement '%s' is the name of a collection" % collection)
```
Nitpick: inconsistent interpolation style. Sometimes it's `str.format()` and sometimes it's `%`. It's better to stick with one option everywhere.

s-hertel replied:
Fixed.

webknjaz commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1216,3 +1422,13 @@ def _consume_file(read_from, write_to=None):
+
+def get_galaxy_metadata_path(b_path):
+    b_default_path = os.path.join(b_path, b'galaxy.yml')
+    candidate_names = [b'galaxy.yml', b'galaxy.yaml']
```
You have only one variant in many places hardcoded. I'm expecting this to cause bugs... How about having at least a module-global constant for these? Or maybe a structure (namedtuple?) that holds both filename and its contents.

s-hertel replied:
@webknjaz Where is that? Can you give me an example? We've only supported `galaxy.yml` in the past. If this looks invasive I'll just remove this supporting both for now, and it can be addressed in a follow-up PR.

webknjaz replied:
Line 1211

s-hertel replied:
Fixed, thanks

webknjaz commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,84 @@
+You can install a collection in a git repository by providing the URI to the repository instead of a collection name or path to a ``tar.gz`` file. The collection must contain a ``galaxy.yml`` file, which will be used to generate the would-be collection artifact data from the directory. The URI should be prefixed with ``git+`` (or with ``git@`` to use a private repository with ssh authentication)  ...
```
URL and URI seem to be used inconsistently across the docs.

P.S. Nitpick: it's better to split paragraphs into multiple shorter lines in order to make the rst source more readable.

s-hertel replied:
Roles support URLs. Collections support URIs.

webknjaz replied:
I think I saw mentions of URLs below in this document

s-hertel replied:
Fixed

webknjaz commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,84 @@
+
+.. warning::
+
+   Use SSH keys instead of embedding credentials into a git URL.
```
AFAIR one may put creds (access token) into `~/.netrc` and it'd work just fine.

s-hertel replied:
Ok, I can reword that to just say not to embed credentials instead of listing secure options.

webknjaz replied:
Well, removing isn't necessary but you could add something like "Make sure to use safe auth options for security reasons. For example, use SSH, netrc or [`http.extraHeader`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-httpextraHeader)/[`url.<base>.pushInsteadOf`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-urlltbasegtinsteadOf) in Git config to prevent your creds from being exposed in logs."

s-hertel replied:
Fixed

webknjaz commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,84 @@
+.. code-block:: text
+
+   ├── galaxy.yml
+   ├── plugins
+   │   ├── lookup
+   │   ├── modules
+   │   └── module_utils
```
I'd mark folders `ls -F` style:
```suggestion
   ├── plugins/
   │   ├── lookup/
   │   ├── modules/
   │   └── module_utils/
```

webknjaz commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,84 @@
+
+.. code-block:: text
+
+   directory
+   ├── docs
+   ├── galaxy.yml
+   ├── plugins
+   │   ├── inventory
+   │   └── modules
+   └── roles
```
```suggestion
   directory/
   ├── docs/
   ├── galaxy.yml
   ├── plugins/
   │   ├── inventory/
   │   └── modules/
   └── roles/
```

webknjaz commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,84 @@
+
+.. code-block:: text
+
+   namespace
+   └── name
+       ├── docs
+       ├── galaxy.yml
+       ├── plugins
+       │   ├── README.md
+       │   └── modules
+       ├── README.md
+       └── roles
```
```suggestion
   namespace/
   └── name/
       ├── docs/
       ├── galaxy.yml
       ├── plugins/
       │   ├── README.md
       │   └── modules/
       ├── README.md
       └── roles/
```

webknjaz commented on docs/docsite/rst/shared_snippets/installing_collections_git_repo.txt:
```diff
@@ -0,0 +1,84 @@
+.. code-block:: bash
+
+   # Install all collections in a particular namespace
+   ansible-galaxy collection install git+https://github.com/organization/repo_name.git#/namespace/
```
Is trailing splash mandatory? If yes — document this.

s-hertel replied:
It works the same with and without the trailing slash.

webknjaz commented on test/integration/targets/ansible-galaxy-collection-scm/aliases:
```diff
@@ -0,0 +1,3 @@
+shippable/posix/group4
+skip/aix
+skip/python2.6  # ansible-galaxy uses tarfile with features not available until 2.7
```
Also, controller dropped python 2.6 a while back anyway

webknjaz commented on test/integration/targets/ansible-galaxy-collection-scm/tasks/scm_dependency_deduplication.yml:
```diff
@@ -0,0 +1,50 @@
+
+- assert:
+    that:
+      - "'namespace_1.collection_1' in installed_collections.stdout"
+      - "'namespace_2.collection_2' in installed_collections.stdout"
```
Pro tip: you can often get rid of escaping when using YAML blocks:
```suggestion
      - >-
        'namespace_1.collection_1' in installed_collections.stdout
      - >-
        'namespace_2.collection_2' in installed_collections.stdout
```

s-hertel replied:
:+1: I like the escaping better than the extra lines though in this case.

s-hertel committed:
Update with webknjaz's suggestions

webknjaz commented:
That linter error should probably be suppressed because this module runs on the controller AFAIU

s-hertel committed:
suppress sanity compile failure for Python 2.6

add links for auth options to the docs

s-hertel committed:
Update with shrews's suggestion to add the shared snippet to user_guide/collection_using.rst
